### ISSUE TITLE
POWER: give power blocks particle collision functionality

### ISSUE THREAD

**Issue by lukephills** [BlokDust/BlokDust#171]
I'd like to add a particle collision function to Power blocks. This would allow the power to be reset on hit so that any sources attached to it would restart from triggerAttack position. 

Example:
A Power connected to a Soundcloud (which has loop turned off). Currently the sample plays through until the waveform end point. Shooting a particle at the power would instantly restart the loop from the start position. 

This would help scheduling sample playback for predetermined lengths.

**Comment by edsilv** [comment]
My intuition about powering a soundcloud block would be that while powered, it continues playing until the end, or plays again from the start if looping is enabled. Removing the power resets it to the start. Perhaps whether it resets to start when unpowered is a block setting.

So we'd want to quickly turn the power off and on again to reset it - which is what you're saying I think.

What if a void/barrier could be powered? A rotating laser or particle emitter could supply a momentary burst of power to it, which could then block a laser powering a soundcloud block for an instant, thus resetting it.

**Comment by lukephills** [comment]
> So we'd want to quickly turn the power off and on again to reset it - which is what you're saying I think.

Yep that's right. I'm seeing it like a particle or laser hits a power block and gives it a extra burst of energy. The sources that are connected are all still active but the attack portion of their envelopes would be retriggered. 

I'm not sure powering void's would work since they already 'drain' power.

**Comment by Whitevinyl** [comment]
I feel there's another way we could do it, not sure how yet though.
Power blocks in my mind should remain the dumbest thing with no functionality, they output power consistently and that's all we need to know about them. What's good about them is that they don't interfere with particles & lasers.

I'd like to extend the logic to have a "not" power at some point. It could be powering a Soundcloud block, but when hit by a particle it has a moment (our pulse length) of not powering, before resuming, and retriggering any sources attached. adding an envelope with release to the Soundcloud block would smooth over the gap if that's an issue.

Part of my refactoring power plan is that I don't feel any power or logic behaviour should be directly calling attack or release methods on the audio.
Instead we'd have AddPower() and RemovePower() functions in Source. In those functions, when a source goes from having zero power connections to more than zero, then the Attack is triggered, and when going from more than zero power connections to zero, release is triggered. That way multiple powers aren't competing to trigger audio and we should have to have different cases for different types of block.
